pkg/cachemanager/cachemanager_test/cachemanager_integration_test.go

```diff
@@ -0,0 +1,333 @@
+		},
+	}
+
+	cacheManager, dataStore, _, ctx := cacheManagerForTest(t, mgr, wm, reader)
+
+	cfClient, ok := dataStore.(*cachemanager.FakeCfClient)
+	require.True(t, ok)
+
+	// seed one gvk
+	configMapGVK := schema.GroupVersionKind{Group: "", Version: "v1", Kind: "ConfigMap"}
+	cm := unstructuredFor(configMapGVK, "config-test-1")
+	require.NoError(t, c.Create(ctx, cm), "creating ConfigMap config-test-1")
+
+	podGVK := schema.GroupVersionKind{Group: "", Version: "v1", Kind: "Pod"}
+	pod := unstructuredFor(podGVK, "pod-1")
+	require.NoError(t, c.Create(ctx, pod), "creating Pod pod-1")
+
+	syncSourceOne := aggregator.Key{Source: "source_a", ID: "ID_a"}
+	require.NoError(t, cacheManager.UpsertSource(ctx, syncSourceOne, []schema.GroupVersionKind{configMapGVK, podGVK}))
+
+	expected := map[cachemanager.CfDataKey]interface{}{
+		{Gvk: configMapGVK, Key: "default/config-test-1"}: nil,
+		{Gvk: podGVK, Key: "default/pod-1"}:               nil,
+	}
+
+	require.Eventually(t, expectedCheck(cfClient, expected), eventuallyTimeout, eventuallyTicker)
+
+	// set up a scenario where the list from replay will fail a few times
+	failPlease <- "ConfigMapList"
+	failPlease <- "ConfigMapList"
+
+	// this call should schedule a cache wipe and a replay for the configMapGVK
+	require.NoError(t, cacheManager.UpsertSource(ctx, syncSourceOne, []schema.GroupVersionKind{configMapGVK}))
+
+	expected2 := map[cachemanager.CfDataKey]interface{}{
+		{Gvk: configMapGVK, Key: "default/config-test-1"}: nil,
+	}
+	require.Eventually(t, expectedCheck(cfClient, expected2), eventuallyTimeout, eventuallyTicker)
+
+	// cleanup
+	require.NoError(t, c.Delete(ctx, cm), "creating ConfigMap config-test-1")
+	require.NoError(t, c.Delete(ctx, pod), "creating ConfigMap pod-1")
+}
+
+// TestCacheManager_concurrent makes sure that we can add and remove multiple sources
+// from separate go routines and changes to the underlying cache are reflected.
+func TestCacheManager_concurrent(t *testing.T) {
+	mgr, wm := testutils.SetupManager(t, cfg)
+	c := testclient.NewRetryClient(mgr.GetClient())
+	cacheManager, dataStore, agg, ctx := cacheManagerForTest(t, mgr, wm, c)
+
+	configMapGVK := schema.GroupVersionKind{Group: "", Version: "v1", Kind: "ConfigMap"}
+	podGVK := schema.GroupVersionKind{Group: "", Version: "v1", Kind: "Pod"}
+
+	// Create configMaps to test for
+	cm := unstructuredFor(configMapGVK, "config-test-1")
+	require.NoError(t, c.Create(ctx, cm), "creating ConfigMap config-test-1")
+
+	cm2 := unstructuredFor(configMapGVK, "config-test-2")
+	require.NoError(t, c.Create(ctx, cm2), "creating ConfigMap config-test-2")
+
+	pod := unstructuredFor(podGVK, "pod-1")
+	require.NoError(t, c.Create(ctx, pod), "creating Pod pod-1")
+
+	cfClient, ok := dataStore.(*cachemanager.FakeCfClient)
+	require.True(t, ok)
+
+	syncSourceOne := aggregator.Key{Source: "source_a", ID: "ID_a"}
+	syncSourceTwo := aggregator.Key{Source: "source_b", ID: "ID_b"}
+
+	wg := &sync.WaitGroup{}
+
+	wg.Add(2)
+	go func() {
```

Comment:
I wonder if it would be better to have this section do something like loop while creating goroutines to approximate a lot of load?

```
wg := &sync.WaitGroup{}
for i := 1; i < 100; i++ {
  go func() {
    wg.Add(1)
    defer wg.Done()
    // require 1
  }()
  go func() {
    wg.Add(1)
    defer wg.Done()
    // require 2
  }()
}
wg.Wait()
```

Reply:
+1, additions and removals happening at random times, along with the adding/deletion of objects-to-be-synced are more likely to flush out deadlocks.

Hardest bit will be knowing the system converges on a known final state. Suggest:

1) a go routine that alternately adds and removes config map syncs, with some random jitter. The last command is add.

2) a go routine that alternately adds and removes pod syncs, with some random jitter. The last command is add.

3) a config map that is always present + a goroutine that creates/deletes a config map (last command is create).

Let these three go routines run, then make sure they converge on expected state.

This could also take the place of the section:

> // now swap the gvks for each source and do so repeatedly to generate some churn

The "// now remove the podgvk for sync source two and make sure we don't have pods in the cache anymore"

Section is good to keep after doing the above to make sure replays haven't been affected by a deadlock.

Lastly, this code is bulky enough that I'd break it up into subtests (they'd still all live under the one larger function). Example:

```golang
func MyTest(t *testing.T) {
   // create global dependencies, such as client and cache manager
   t.Run("MySubtest", func(t *testing.T) {
      // do some things
      // make assertions
   }
   t.Run("MySecondSubtest", func(t *testing.T {
      // do more things
      // make assertions
   }
}
```

Reply:
Thanks for chiming in Max! 

> a go routine that alternately adds and removes config map syncs, with some random jitter. The last command is add.
> a go routine that alternately adds and removes pod syncs, with some random jitter. The last command is add.

I think we have this setup a bit currently by virtue of the upserts, less the jitter.

is the jitter important? if so, can it be added as a follow up as part of ? https://github.com/open-policy-agent/gatekeeper/issues/2929

Regarding

> a config map that is always present + a goroutine that creates/deletes a config map (last command is create).

we have a instance update test for the sync controller. Is there value in combining that with the sync source updates?

finally regarding using substeps, I can try to do that in a follow up PR. My experience has been that subtests are easy to mess up in situations like these because of any shared state

Reply:
Following up per offline chat, the latest commits add jitter inside the go routine calls and I made this test shorter.